Source/WebKit/UIProcess/API/gtk/DropTargetGtk4.cpp

```diff
@@ -129,20 +132,56 @@ void DropTarget::accept(GdkDrop* drop, std::optional<WebCore::IntPoint> position
         }, this);
     }
 
+    static const char* const portalMIMETypes[] = {
+        "application/vnd.portal.filetransfer",
+        "application/vnd.portal.files", // Deprecated, but added for compatibility
+    };
+
     static const char* const supportedMimeTypes[] = {
+        "application/vnd.portal.filetransfer",
+        "application/vnd.portal.files", // Deprecated, but added for compatibility
         "text/html",
         "_NETSCAPE_URL",
         "text/uri-list",
         "application/vnd.webkitgtk.smartpaste",
         "org.webkitgtk.WebKit.custom-pasteboard-data"
     };
 
+    bool transferredFilesFromPortal = false;
     for (unsigned i = 0; i < G_N_ELEMENTS(supportedMimeTypes); ++i) {
         if (!gdk_content_formats_contain_mime_type(formats, supportedMimeTypes[i]))
             continue;
 
+        // Reading from the File Transfer portal is a bit special. When either portal
+        // mimetypes are present, GTK serializes them using the GdkFileList type. If
+        // this type is present, ignore file:// URIs from the "text/uri-list" later on.
+        if (!transferredFilesFromPortal && g_strv_contains(portalMIMETypes, supportedMimeTypes[i])) {
+            ASSERT(gdk_content_formats_contain_gtype(formats, GDK_TYPE_FILE_LIST));
+
+            m_dataRequestCount++;
+            gdk_drop_read_value_async(m_drop.get(), GDK_TYPE_FILE_LIST, G_PRIORITY_DEFAULT, m_cancellable.get(), [](GObject* gdkDrop, GAsyncResult* result, gpointer userData) {
+                GUniqueOutPtr<GError> error;
+                const GValue* value = gdk_drop_read_value_finish(GDK_DROP(gdkDrop), result, &error.outPtr());
+                if (g_error_matches(error.get(), G_IO_ERROR, G_IO_ERROR_CANCELLED))
+                    return;
+
+                auto& drop = *static_cast<DropTarget*>(userData);
+                if (value && G_VALUE_HOLDS(value, GDK_TYPE_FILE_LIST)) {
+                    StringBuilder uriListBuilder;
+                    GSList* files = static_cast<GSList*>(g_value_get_boxed(value));
+                    for (GSList *l = files; l; l = l->next)
+                        drop.m_transferredFiles.append(G_FILE(l->data));
+                }
+
+                drop.didLoadData();
```

Comment:
It's a bit weird to call `didLoadData()` when `loadData()` hasn't been called. Maybe we can add another loadData that receives a `CompletionHandler<Vector<GRefPtr<GFile>>>&&`